### PR TITLE
Added code for marking not updated nodes

### LINKED ISSUE #107973: Bug in node_status_updater.go - UpdateNodeStatuses
https://github.com/kubernetes/kubernetes/blob/9b09612d1b0a6ec441833f928cbb205d936da59c/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L92

UpdateNodeStatuses will exit immediately if updateNodeStatus returns an error.
It does mark the node for which the call failed for later update but all the following nodes in the update list are left unchanged and therefore won't be updated.

This can result for instance in a pod getting stuck waiting for its volume to be attached.
Assuming the attach of the volume to the pod's node succeeded but another node preceding this one in the update list has its updateNodeStatus call fail - then the attach detach controller will never update the node status to indicate the volume was attached and the pod will be stuck until there is another update for this node attachements.

The fix can be as simple as using a continue but the original code seems to want to abort the whole loop in the case of that error, so that would require marking all the remaining nodes for update, not just the current one.

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
This PR remarks remaining nodes in the nodeList, informing that this particular node needs status update. 
#### Which issue(s) this PR fixes:

Fixes #107973

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go [hunk 1]
```diff
@@ -87,7 +89,16 @@ func (nsu *nodeStatusUpdater) UpdateNodeStatuses() error {
...
-
+			failedNodeNumber := nodeNumber
+			var i int
+			//re-marking remaining nodes for update
+			for nodeName, _ = range nodesToUpdate {
+				i++
+				if i < failedNodeNumber+1 {
+					continue
+				}
+				nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
```
Comment (NikhilSharmaWe) [inline pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:100]: Should I add log statement for these nodes also similar to [this](https://github.com/kubernetes/kubernetes/pull/107976/files#diff-4505a6c5e6e8d09f1f5ff559a9ca14e4efa62131d211ea151f2f86dc295ca982R70-R73) statement which was added for the failed node .

Comment (jfremy) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:100]: This logic might be dangerous. I don't think there is any guarantee in the go language that iterating over a map twice will result in the same iteration order (see for instance https://go.dev/blog/maps - Iteration order)
So if we want to  proceed with aborting the call altogether if there is an error in that clause, we might want to introduce an abort boolean in the main scope and set it to true if there is such an error and have a check at the beginning of the loop that will check for abort and if true will call nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded and continue to the next entry

Comment (NikhilSharmaWe) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:100]: @jfremy When we will change the value of `var abourt bool` inside the condition `if err := nsu.updateNodeStatus(nodeName, nodeObj, attachedVolumes); err != nil {)` then how will be able to use it to re-mark the remaining nodes because at the end of this condition we `return err` which will end the `function`, so how can we check the value of `abort` for the remaining nodes for re-marking them.

Comment (jfremy) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:100]: The return err needs to be removed in that scenario

Comment (NikhilSharmaWe) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:100]: So there is no need to return this error `err := nsu.updateNodeStatus(nodeName, nodeObj, attachedVolumes); err != nil` in the function.

Comment (jfremy) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go]: The beginning of the loop would have the abort condition
```
if abort {
  nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
  continue
}
```

while the error path for updateNodeStatus would look like this:
```
if err := nsu.updateNodeStatus(nodeName, nodeObj, attachedVolumes); err != nil {
    // If update node status fails, reset flag statusUpdateNeeded back to true
    // to indicate this node status needs to be updated again
    nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
    klog.V(2).Infof(
        "Could not update node status for %q; re-marking for update. %v",
        nodeName,
        err)`
    abort = true
    continue
}
```
with abort being defined as a bool in the outer scope of the function

But again, that's only if we want to go in the abort the whole call direction. I don't have the context of why the original code writer made that choice. Maybe replacing `return err` with `continue` would be ok too

Comment (NikhilSharmaWe) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go]: Done, thanks for the help

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go [hunk 2]
```diff
@@ -56,11 +56,17 @@ type nodeStatusUpdater struct {
 	actualStateOfWorld cache.ActualStateOfWorld
 }
 
+var abort bool
```
[the remark this answers is not in the document]
Comment (NikhilSharmaWe) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go]: Fixed the scope

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go (hunk 1 above)
Comment (jingxu97) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go]: abort looks not very necessary here? Basically it needs iterate the whole node list, and update node status. If any node status update failed, mark nodeStatusUpdateNeed in actual state of world, and continue to the next item on the list. The return error in this function is not very necessary either, it is right now just for logging.

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go [hunk 3]
```diff
@@ -57,10 +57,15 @@ type nodeStatusUpdater struct {
 }
 
 func (nsu *nodeStatusUpdater) UpdateNodeStatuses() error {
+	var abort bool
```
Comment (jingxu97) [inline pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:60]: just to be extra safe maybe directly setting to false
abort := false

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go (hunk 1 above)
Comment (jingxu97) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go]: There are different possible node update failures. 
1. In case of somehow node does not exist, node update failed. In this case, we want to continue updating other nodes.
2. In case API server issue, all update might fail. In this case, abort the rest operation is a good choice to reduce loading api more.
But I think to be safer, we should try to continue to update the rest of the nodes in the list. Otherwise, one node update failure will always block the rest of the list.  

Basically it iterates the whole node list, and update node status. If any node status update failed, mark nodeStatusUpdateNeed in actual state of world, and continue to the next item on the list. The return error in this function is not very necessary either.

**Comment by jingxu97** [comment]
There are different possible node update failures. 
1. In case of somehow node does not exist, node update failed. In this case, we want to continue updating other nodes.
2. In case API server issue, all update might fail. In this case, abort the rest operation is a good choice to reduce loading api server more.

But I think to be safer, we should try to continue to update the rest of the nodes in the list. Otherwise, one node update failure will always block the rest of the nodes on the list.  

Basically it iterates the whole node list, and update node status. If any node status update failed, mark nodeStatusUpdateNeed in actual state of world, and continue to the next item on the list. The return error in this function is not very necessary either.

**Comment by NikhilSharmaWe** [comment]
@jingxu97 

- We want to update the rest of the nodes in both case (when node is not found, when API server issue).
- So the only change we need in this PR is to replace [`return err`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L92) with `continue` in the condition :
 `if err := nsu.updateNodeStatus(nodeName, nodeObj, attachedVolumes); err != nil{}`

Could you please inform that am I understanding right here or not ?

**Comment by gnufied** [comment]
> The return error in this function is not very necessary either.

@jingxu97 the return error is necessary in case , if we want to make sure it is safe to detach volume from the node. If this function returned no error then - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L205 will try to detach volume from a node, without potentially marking the node.

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go [hunk 4]
```diff
@@ -57,10 +57,15 @@ type nodeStatusUpdater struct {
...
 	// TODO: investigate right behavior if nodeName is empty
 	// kubernetes/kubernetes/issues/37777
 	nodesToUpdate := nsu.actualStateOfWorld.GetVolumesToReportAttached()
 	for nodeName, attachedVolumes := range nodesToUpdate {
+		if abort {
+			nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:68]: On second thought - aggregating error from all nodes may be problematic because we will be constructing these huge error objects which might not be very nice for memory pressure (until they get GCed). So, I would not perhaps use same code. 

But we *do* need to return an error if we can't update nodes.

[the remark this answers is not in the document]
Comment (NikhilSharmaWe) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:68]: @jingxu97, @jfremy Could someone give their thoughts on the changes suggested by @gnufied. In which direction should we go regarding solving this bug if we have to ?

**Comment by YuikoTakada** [comment]
I suggest:
* Add UpdateNodeStatus() to node_status_updater.go. UpdateNodeStatus() updates the status of just one node, not all nodes.
```
func (nsu *nodeStatusUpdater) UpdateNodeStatus(nodeName) error {
...
		if err := nsu.updateNodeStatus(nodeName, nodeObj, attachedVolumes); err != nil {
		    return err
}
```
* Call UpdateNodeStatus() from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L205
```
			// Update Node Status to indicate volume is no longer safe to mount.
			err = rc.nodeStatusUpdater.UpdateNodeStatus(attachedVolume.NodeName)
			if err != nil {
				// Skip detaching this volume if unable to update node status
				klog.ErrorS(err, "UpdateNodeStatuses failed while attempting to report volume as attached", "volume", attachedVolume)
				continue
			}
```
* UpdateNodeStatuses() seems to be called just by https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L205 and https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L243 , so replace former with UpdateNodeStatus() and later would be good as it is.

**Comment by jfremy** [comment]
I took an approach similar to what @YuikoTakada suggested. And also fixed what I think was another bug in the process.

**Comment by jingxu97** [comment]
we can close this one
